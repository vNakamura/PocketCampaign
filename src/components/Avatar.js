import React, { PropTypes } from 'react';
import isNil from 'lodash/isNil';
import md5 from 'md5';

import style from './Avatar.styl';

function getImageSrc(props) {
  const { name, image } = props;

  if (isNil(image)) {
    return `https://api.adorable.io/avatars/128/${md5(name)}`;
  }
  return image;
}

const Avatar = (props) =>
  <div className={props.className}><img
    className={style.avatar}
    title={props.name}
    role="presentation"
    src={getImageSrc(props)}
    style={props.style}
  /></div>;

Avatar.propTypes = {
  name: PropTypes.string.isRequired,
  image: PropTypes.string,
  style: PropTypes.object,
  className: PropTypes.string,
};

export default Avatar;
