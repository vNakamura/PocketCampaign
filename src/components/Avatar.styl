@import "../config.styl"

.avatar
  border-radius: 50%;
  border: solid 4px $canvas;
  box-shadow: -10px 10px 32px rgba(0,0,0,0.1);
