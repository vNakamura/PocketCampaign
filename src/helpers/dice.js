// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
