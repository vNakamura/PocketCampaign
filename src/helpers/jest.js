import React from 'react';
import { ThemeProvider } from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import 'jest-styled-components';
/* eslint-enable import/no-extraneous-dependencies */
import theme from '../theme';

export const applyTheme = content => (
  <ThemeProvider theme={theme}>
    {content}
  </ThemeProvider>
);

export const renderWithTheme = content => renderer.create(
  applyTheme(content),
).toJSON();

export default {
  applyTheme,
  renderWithTheme,
};
