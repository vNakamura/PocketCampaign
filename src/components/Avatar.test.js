import React from 'react';
import { mount } from 'enzyme';

import { renderWithTheme } from '../helpers/jest';
import Avatar from './Avatar';

describe('components: Avatar', () => {
  it('should render correctly', () => {
    const avatar = renderWithTheme(<Avatar user={{ key: 'someKey' }} />);
    expect(avatar).toMatchSnapshot();
  });
  it("should use a placeholder image if user doesn't have one", () => {
    const wrapper = mount(<Avatar user={{ avatar: 'avatarPath' }} />);
    expect(wrapper.find('img').prop('src')).toEqual('avatarPath');
    wrapper.setProps({
      user: { key: 'someKey' },
    });
    const src = wrapper.find('img').prop('src');
    expect(src).toEqual(expect.stringContaining('http'));
    expect(src).toEqual(expect.stringContaining('someKey'));
  });
});
