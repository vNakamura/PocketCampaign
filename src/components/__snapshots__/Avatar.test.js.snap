// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components: Avatar should render correctly 1`] = `
.c0 {
  width: 10vmin;
  max-width: 50px;
  height: 10vmin;
  max-height: 50px;
  border-radius: 50%;
  margin: .4rem;
}

<img
  className="c0"
  src="https://api.adorable.io/avatars/128/someKey"
/>
`;
